### PR TITLE
[language] Add transaction fee distribution to the block prologue

### PR DESCRIPTION
**Pull request by tzakian**
Adds transaction fee distribution to the libra system module and calls it in the block prologue (for now). Once we have an on-chain representation of epochs then we can update this to only happen once per-epoch[1].

[1] I considered putting this in the `reconfigure` function, but this seems to also be called per-block, and we want transaction fees to be paid out even if no validators may have changed, so I decided to not call this from within that function.

### CODE REVIEW

File: language/functional_tests/tests/testsuite/validator_set/reconfiguration.mvir [hunk 1]
```diff
@@ -83,7 +82,7 @@ main() {
     return;
 }
 
-// check: events: []
+// not: ValidatorSetChangeEvent
```
Comment (runtian-zhou) [inline language/functional_tests/tests/testsuite/validator_set/reconfiguration.mvir]: Should we also check the balance of the validators to make sure that gas are indeed distributed across the network?

File: language/stdlib/modules/libra_system.mvir [hunk 2]
```diff
@@ -311,4 +320,121 @@ module LibraSystem {
...
+    //    TODO: Once we have a better on-chain representation of
+    //          epochs this should be changed over to be once per-epoch.
+    //  * In the handling of dust, we simply leave it to the next distribution.
+    //    This is meant to in part minimize the benefit of being the first validator in
+    //    the validator set.
+
+     // Initialize the transaction fee distribution module. We keep track of the last paid block
+     // height in order to ensure that we don't try to pay more than once per-block. We also
+     // encapsulate the withdrawal capability to the transaction fee account so that we can withdraw
+     // the fees from this account from block metadata transactions.
```
Comment (sblackshear) [inline language/stdlib/modules/libra_system.mvir]: Nice, good use-case for the `WithdrawalCapability`

File: language/functional_tests/tests/testsuite/validator_set/reconfiguration.mvir (hunk 1 above)
Comment (sblackshear) [reply language/functional_tests/tests/testsuite/validator_set/reconfiguration.mvir]: Agreed; I think we should at least create versions of the deleted `test_txn_fee_distribution_does_div.mvir` and `test_txn_fee_distribution_no_div.mvir` files that test the same thing in this new setting.

**Comment by tzakian** [comment]
Ahh good point on those tests! I'll add those back in.

File: language/functional_tests/tests/testsuite/validator_set/reconfiguration.mvir (hunk 1 above)
Comment (tzakian) [reply language/functional_tests/tests/testsuite/validator_set/reconfiguration.mvir]: I've added those two deleted tests back, but not in this file.

File: language/functional_tests/tests/testsuite/transaction_fee_distribution/test_txn_fee_distribution_does_div.mvir [hunk 3]
```diff
@@ -1,27 +1,52 @@
 // Number of validators evenly divides the transaction fees collected
 
+//! account: vivian, 1000000, 0, validator
+//! account: bobby, 1000000, 0, validator
```
Comment (sblackshear) [inline language/functional_tests/tests/testsuite/transaction_fee_distribution/test_txn_fee_distribution_does_div.mvir]: Come on man, let's see some creative "V" names :)

File: language/functional_tests/tests/testsuite/transaction_fee_distribution/test_txn_fee_distribution_no_div.mvir [hunk 4]
```diff
@@ -1,37 +1,61 @@
...
-       assert(LibraAccount.balance(move(addr)) == 10, 3);
-     }
+     new_balance = LibraAccount.balance(move(addr));
+     old_balance = Vector.get<u64>(&balances, copy(index));
      index = move(index) + 1;
+     assert(move(new_balance) == (move(old_balance) + 10), 1);
   }
 
-  assert(LibraAccount.balance({{fees}}) < move(number_of_validators), 4);
+  // The remainder in the 0XFEE's must be less than the number of validators.
```
Comment (sblackshear) [inline language/functional_tests/tests/testsuite/transaction_fee_distribution/test_txn_fee_distribution_no_div.mvir]:
```suggestion
  // The remainder in 0XFEE's balance must be less than the number of validators.
```

File: language/functional_tests/tests/testsuite/transaction_fee_distribution/test_txn_fee_distribution_does_div.mvir (hunk 3 above)
Comment (tzakian) [reply language/functional_tests/tests/testsuite/transaction_fee_distribution/test_txn_fee_distribution_does_div.mvir]: Fixed! Had to google baby names though lol

**Pull request closed without merging**